File: services/academic-service/src/test/java/com/visor/school/academicservice/service/StudentServiceTest.java
package com.visor.school.academicservice.service;

import com.visor.school.academicservice.event.StudentEventPublisher;
import com.visor.school.academicservice.model.EnrollmentStatus;
import com.visor.school.academicservice.model.Student;
import com.visor.school.academicservice.repository.StudentRepository;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class StudentServiceTest {

    @Mock
    private StudentRepository studentRepository;

    @Mock
    private StudentEventPublisher studentEventPublisher;

    @Mock
    private StudentIdGenerator studentIdGenerator;

    @InjectMocks
    private StudentService studentService;

    private final Student testStudent = new Student(
            "S12345",
            UUID.randomUUID(),
            "John",
            "Doe",
            LocalDate.of(2010, 5, 15),
            5,
            EnrollmentStatus.ENROLLED,
            null,
            null
    );

    @Test
    void enrollStudentShouldCreateAndPublishEvent() {
        UUID userId = testStudent.getUserId();
        Student savedStudent = new Student(
                testStudent.getStudentId(),
                userId,
                testStudent.getFirstName(),
                testStudent.getLastName(),
                testStudent.getDateOfBirth(),
                testStudent.getGradeLevel(),
                testStudent.getEnrollmentStatus(),
                null,
                null
        );
        savedStudent.setId(UUID.randomUUID());

        when(studentIdGenerator.generateStudentId()).thenReturn("S12345");
        when(studentRepository.save(any(Student.class))).thenReturn(savedStudent);

        Student enrolledStudent = studentService.enrollStudent(
                userId,
                testStudent.getFirstName(),
                testStudent.getLastName(),
                testStudent.getDateOfBirth(),
                testStudent.getGradeLevel(),
                null,
                null
        );

        assertNotNull(enrolledStudent.getStudentId());
        verify(studentRepository).save(any(Student.class));
        verify(studentEventPublisher).publishStudentEnrolled(enrolledStudent);
    }

    @Test
    void updateStudentShouldSaveAndPublishEvent() {
        UUID studentId = UUID.randomUUID();
        Student existingStudent = new Student(
                testStudent.getStudentId(),
                testStudent.getUserId(),
                testStudent.getFirstName(),
                testStudent.getLastName(),
                testStudent.getDateOfBirth(),
                testStudent.getGradeLevel(),
                testStudent.getEnrollmentStatus(),
                null,
                null
        );
        existingStudent.setId(studentId);

        Student updatedStudent = new Student(
                testStudent.getStudentId(),
                testStudent.getUserId(),
                "Jane",
                testStudent.getLastName(),
                testStudent.getDateOfBirth(),
                testStudent.getGradeLevel(),
                testStudent.getEnrollmentStatus(),
                null,
                null
        );
        updatedStudent.setId(studentId);

        when(studentRepository.findById(studentId)).thenReturn(Optional.of(existingStudent));
        when(studentRepository.save(any(Student.class))).thenReturn(updatedStudent);

        Student result = studentService.updateStudent(studentId, "Jane", null, null, null, null);

        assertEquals("Jane", result.getFirstName());
        verify(studentRepository).save(any(Student.class));
    }

    @Test
    void searchStudentsShouldReturnAPageOfStudents() {
        when(studentRepository.searchByName("John")).thenReturn(List.of(testStudent));

        List<Student> students = studentService.searchStudentsByName("John");

        assertFalse(students.isEmpty());
        assertEquals(1, students.size());
        verify(studentRepository).searchByName("John");
    }

    @Test
    void getStudentsByGradeShouldReturnAList() {
        when(studentRepository.findByGradeLevel(5)).thenReturn(List.of(testStudent));

        List<Student> students = studentService.getStudentsByGradeLevel(5);

        assertFalse(students.isEmpty());
        assertEquals(1, students.size());
        verify(studentRepository).findByGradeLevel(5);
    }
}

File: services/academic-service/src/test/kotlin/com/visor/school/academicservice/service/StudentServiceTest.kt
package com.visor.school.academicservice.service

import com.visor.school.academicservice.event.StudentEventPublisher
import com.visor.school.academicservice.model.EnrollmentStatus
import com.visor.school.academicservice.model.Student
import com.visor.school.academicservice.repository.StudentRepository
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.ArgumentMatchers.anyString
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.kotlin.*
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.PageRequest
import java.time.LocalDate
import java.util.Optional
import java.util.UUID

@ExtendWith(MockitoExtension::class)
class StudentServiceTest {

    @Mock
    private lateinit var studentRepository: StudentRepository

    @Mock
    private lateinit var studentEventPublisher: StudentEventPublisher

    @Mock
    private lateinit var studentIdGenerator: StudentIdGenerator

    @InjectMocks
    private lateinit var studentService: StudentService

    private val testStudent = Student(
        userId = UUID.randomUUID(),
        studentId = "S12345",
        firstName = "John",
        lastName = "Doe",
        dateOfBirth = LocalDate.of(2010, 5, 15),
        gradeLevel = 5,
        enrollmentStatus = EnrollmentStatus.ENROLLED
    )

    @Test
    fun `enroll student should create and publish event`() {
        // Given
        val savedStudent = Student(
            id = UUID.randomUUID(),
            userId = testStudent.userId,
            studentId = testStudent.studentId,
            firstName = testStudent.firstName,
            lastName = testStudent.lastName,
            dateOfBirth = testStudent.dateOfBirth,
            gradeLevel = testStudent.gradeLevel,
            enrollmentStatus = testStudent.enrollmentStatus
        )
        whenever(studentIdGenerator.generateStudentId()).thenReturn("S12345")
        whenever(studentRepository.save(any<Student>())).thenReturn(savedStudent)

        // When
        val enrolledStudent = studentService.enrollStudent(
            testStudent.userId,
            testStudent.firstName,
            testStudent.lastName,
            testStudent.dateOfBirth,
            testStudent.gradeLevel
        )

        // Then
        assertNotNull(enrolledStudent.studentId)
        verify(studentRepository).save(any<Student>())
        verify(studentEventPublisher).publishStudentEnrolled(enrolledStudent)
    }

    @Test
    fun `update student should save and publish event`() {
        // Given
        val studentId = UUID.randomUUID()
        val existingStudent = Student(
            id = studentId,
            userId = testStudent.userId,
            studentId = testStudent.studentId,
            firstName = testStudent.firstName,
            lastName = testStudent.lastName,
            dateOfBirth = testStudent.dateOfBirth,
            gradeLevel = testStudent.gradeLevel,
            enrollmentStatus = testStudent.enrollmentStatus
        )
        val updatedStudent = Student(
            id = studentId,
            userId = testStudent.userId,
            studentId = testStudent.studentId,
            firstName = "Jane",
            lastName = testStudent.lastName,
            dateOfBirth = testStudent.dateOfBirth,
            gradeLevel = testStudent.gradeLevel,
            enrollmentStatus = testStudent.enrollmentStatus
        )
        whenever(studentRepository.findById(studentId)).thenReturn(Optional.of(existingStudent))
        whenever(studentRepository.save(any<Student>())).thenReturn(updatedStudent)

        // When
        val result = studentService.updateStudent(studentId, "Jane", null, null)

        // Then
        assertEquals("Jane", result.firstName)
        verify(studentRepository).save(any<Student>())
    }

    @Test
    fun `search students should return a page of students`() {
        // Given
        whenever(studentRepository.searchByName("John"))
            .thenReturn(listOf(testStudent))

        // When
        val students = studentService.searchStudentsByName("John")

        // Then
        assertFalse(students.isEmpty())
        assertEquals(1, students.size)
        verify(studentRepository).searchByName("John")
    }

    @Test
    fun `get students by grade should return a list`() {
        // Given
        whenever(studentRepository.findByGradeLevel(5))
            .thenReturn(listOf(testStudent))

        // When
        val students = studentService.getStudentsByGradeLevel(5)

        // Then
        assertFalse(students.isEmpty())
        assertEquals(1, students.size)
        verify(studentRepository).findByGradeLevel(5)
    }
}